pkg/plugin/aggregation/aggregator.go

```diff
@@ -198,21 +212,45 @@ func (a *Aggregator) processResult(result *plugin.Result) error {
 	// Send an event that we got this result even if we get an error, so
 	// that Wait() doesn't hang forever on problems.
 	defer func() {
-		a.Results[result.ExpectedResultID()] = result
+		a.Results[result.Key()] = result
 		a.resultEvents <- result
 	}()
 
 	if err := a.handleResult(result); err != nil {
 		// Drop a breadcrumb so that we reconsider new results from this result.
-		a.FailedResults[result.ExpectedResultID()] = time.Now()
+		a.FailedResults[result.Key()] = time.Now()
 		return &httpError{
 			err:  fmt.Errorf("error handling result %v: %v", resultID, err),
 			code: http.StatusInternalServerError,
 		}
 	}
 
 	// Upon success, we no longer want to keep processing duplicate results.
-	delete(a.FailedResults, result.ExpectedResultID())
+	delete(a.FailedResults, result.Key())
+
+	return nil
+}
+
+// processStatus is the main aggregator logic for handling the status updates from plugins.
+// We first
+func (a *Aggregator) processStatus(status plugin.StatusUpdate) error {
+	a.resultsMutex.Lock()
```

Comment:
I wouldn't jump at the chance to worry about this right now. The other logic that exists for results handling locks/unlocks the mutex at a higher level so that while it is handling a given result (not just checking for duplicates/etc) no other results "cut in line".

If I moved mutex handling down into those low level calls then we lose the ability to do higher level locking without hitting conflicts.